Port python2 scripts to python3

stormi opened the pull request:
I applied futurize, then a manual fix. `check.py` runs successfully (and I made it output the XML files locally so that I could compare them). I also tested with a template containing non-ASCII utf-8 characters. I could not test `create-guest-templates` yet due to an error why using XenAPI from PyPi.

The last two commits fully transition to python3, without python2 support.

stormi committed:
python 2 to 3: run futurize, stages 1 and 2

Stage 1 only converted one print statement so going directly to stage 2.

Files converted using `futurize -wn`.

Signed-off-by: Samuel Verschelde <stormi-xcp@ylix.fr>

stormi committed:
python 2 to 3: update shebangs for python 3

Signed-off-by: Samuel Verschelde <stormi-xcp@ylix.fr>

stormi committed:
python 2 to 3: don't specify an explicit encoding for toxml() calls

An explicit encoding will change the return type for
a byte string rather than a unicode string.

Signed-off-by: Samuel Verschelde <stormi-xcp@ylix.fr>

stormi committed:
python 2 to 3: remove python2 compatibility layer

* Remove dependency on basestring. Check variables against the `str`
  type instead (we don't expect `bytes` to be manipulated).
* Remove imports from __future__, future and builtins

Signed-off-by: Samuel Verschelde <stormi-xcp@ylix.fr>

MarkSymsCtx commented on guesttemplates/blank_template.py:
```diff
@@ -107,7 +107,7 @@ def __init__(self, size, sr, bootable, disk_type):
         doc = minidom.Document()
         entry = doc.createElement('disk')
-        for element_name, element_value in self.__dict__.items():
+        for element_name, element_value in list(self.__dict__.items()):
```
This is most likely unnecessary as the for loop will handle the iterator produced from `.items` quite fine

stormi replied:
Indeed. I left them as strictly equivalent to the python2 behaviour but I can remove them after double checking an iterator / a view is appropriate in this context (same for the next similar comments).

MarkSymsCtx commented on guesttemplates/blank_template.py:
```diff
@@ -271,7 +271,7 @@ def toxml(self, version):
         ver_member.appendChild(struct)
 
-        for n, v in version.items():
+        for n, v in list(version.items()):
```
And this one

MarkSymsCtx commented on guesttemplates/blank_template.py:
```diff
@@ -293,10 +293,10 @@ def toxml(self, version):
         snapshot.appendChild(struct2)
 
-        for n2, v2 in record_dict.items():
+        for n2, v2 in list(record_dict.items()):
```
And this one

rosslagerwall replied:
Same here.

MarkSymsCtx commented on guesttemplates/blank_template.py:
```diff
@@ -326,7 +326,7 @@ def toxml(self, version):
                 struct3 = doc.createElement('struct')
                 value.appendChild(struct3)
-                for n3, v3 in v2.items():
+                for n3, v3 in list(v2.items()):
```
And again

rosslagerwall replied:
And here.

psafont commented:
Could you try generating the XenAPI package straight out from xen-api? this can be done by setting up the xapi ocaml environment and running `make` in  
https://github.com/xapi-project/xen-api/tree/master/scripts/examples/python.
This generates a python wheel which can be installed using pip @benjamreis can help you set up the ocaml environment

rosslagerwall commented on guesttemplates/blank_template.py:
```diff
@@ -271,7 +271,7 @@ def toxml(self, version):
         ver_member.appendChild(struct)
 
-        for n, v in version.items():
+        for n, v in list(version.items()):
```
Why is this needed? `version` is a dict and it should be fine to iterate over the dict_view returned by items() as the loop body doesn't modify the dict.

stormi replied:
Mark commented about the same. It's futurize being conservative by default and converting what used to be a list to a list in python3, just in case we would have used properties of a list that a view doesn't have, such as addressing by index or being a copy you can modify without altering the dict. I'll check that these properties are not used.

rosslagerwall commented on guesttemplates/blank_template.py:
```diff
@@ -107,7 +107,7 @@ def __init__(self, size, sr, bootable, disk_type):
         doc = minidom.Document()
         entry = doc.createElement('disk')
-        for element_name, element_value in self.__dict__.items():
+        for element_name, element_value in list(self.__dict__.items()):
```
Same here.

stormi committed:
python 2 to 3: remove unneeded copy of dict items to lists

`futurize` converts dict.items() found in python2 code to list
in case the code actually requires a list.

Here, a dict_view is appropriate so I'm removing the unneeded
conversion.

Signed-off-by: Samuel Verschelde <stormi-xcp@ylix.fr>

stormi commented:
So, here's what remains to be done according to me:
* Test `create-guest-templates` with python 3. In progress: I will try with a XenAPI python package built from xen-api master branch on @psafont advice.
* Use python-coverage to identify portions of code that may not have been covered by this test or by check.py, and test them manually.

Optionnally, depending on what you ask me:
* Add a github action that executes check.py on push
* Add new test templates, in a directory separate from the `json/` directory. Notably: add non-ascii characters to a template.
* Make check.py write the XML to files. Store them in the repository. Make check.py compare the files it creates with the ones stored. Note: this would require making the output predictable, for example by sorting the nodes (or the use of a smart XML comparison algorithm that is not sensitive to ordering).
* Produce python-coverage output on the tests

andyhhp commented on guesttemplates/blank_template.py:
```diff
@@ -1,3 +1,6 @@
+from builtins import str
+from past.builtins import basestring
+from builtins import object
```
```python
from builtins import object, str
from past.builtins import basestring
```

stormi replied:
I chose (actually, was asked by @edwintorok for this first PR, process to be reviewed afterwards if needed - or so I interpreted it) to first commit exactly what `futurize` produced, as a reference point, and then to improve on it. I could apply this suggestion in a new commit but as all the lines are then removed in the last commit, I'm not sure it's necessary.

stormi replied:
If better for the git history, I can rework it all as 3 commits:
* apply futurize
* fix and complete (squashing all commits that complete or fix what futurize produced)
* remove python2 compatibility

Or squash further and keep only two commits:
* port to python 3, retaining python 2 compatibility
* remove python2 compatibility

Or, which would make things really simpler, just one commit.

andyhhp commented on guesttemplates/blank_template.py:
```diff
@@ -23,7 +20,7 @@ def amount_to_int(amt):
 
 def get_bool_key(d, key, default):
     v = d.get(key, default)
-    if isinstance(v, basestring):
+    if isinstance(v, str):
         v = v.lower() in ('t', 'true', 'y', 'yes', '1')
```
This appears to cause most of the complexity and churn in the port.  How about a starting commit which does:

```python
v = d.get(key, default)
try:
    v = v.lower() in ('t', 'true', 'y', 'yes', '1')
except AttrubteError:
    pass
return v
```

That way, there's no messing around with what is and isn't a string, and works exclusively on whether it has a plausible looking `.lower()` function?

stormi replied:
I like the idea and it would be perfect if it would remove all uses of basestring, but there's a second one for which I think it wouldn't be as elegant once compatibility with python2 is removed. So I suggest keeping the current way: first import `basestring` from `past` as done by `futurize`, then replace it with `str` once we don't need python2 anymore (that is, two commits after).

Or, if we want to make it really simpler, I could now squash everything into one commit and we wouldn't see the somewhat ugly transition that is due to the code being compatible with both python 2 and 3 for a few commits.

stormi commented:
> Could you try generating the XenAPI package straight out from xen-api?

Unfortunately, I still get the same error with the updated XenAPI package (I double checked I was not using the older one):
```
# python3 create-guest-templates 
Traceback (most recent call last):
  File "create-guest-templates", line 10, in <module>
    session.xenapi.login_with_password('', '', '', 'create-template')
  File "/usr/local/lib/python3.6/site-packages/XenAPI/XenAPI.py", line 269, in __call__
    return self.__send(self.__name, args)
  File "/usr/local/lib/python3.6/site-packages/XenAPI/XenAPI.py", line 154, in xenapi_request
    self._login(methodname, params)
  File "/usr/local/lib/python3.6/site-packages/XenAPI/XenAPI.py", line 179, in _login
    getattr(self, 'session.%s' % method)(*params))
  File "/usr/lib64/python3.6/xmlrpc/client.py", line 1112, in __call__
    return self.__send(self.__name, args)
  File "/usr/lib64/python3.6/xmlrpc/client.py", line 1452, in __request
    verbose=self.__verbose
  File "/usr/lib64/python3.6/xmlrpc/client.py", line 1154, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/usr/lib64/python3.6/xmlrpc/client.py", line 1166, in single_request
    http_conn = self.send_request(host, handler, request_body, verbose)
TypeError: send_request() takes 4 positional arguments but 5 were given
```

Maybe there is a bug in the python3 conversion of XenAPI?

psafont commented:
To confirm that the bug only affects python 2, check what happens when running `python2 create-guest-templates`

stormi commented:
> To confirm that the bug only affects python 2, check what happens when running `python2 create-guest-templates`

With python2 it doesn't work yet, but for a different reason: tarfile doesn't like the fact that io.StringIO uses unicode strings in python 2. After solving this, no problem in python2, so the bug above only affects python 3. I'll investigate it.

stormi commented:
Regarding `XenAPI.py`: the bug is likely located at https://github.com/xapi-project/xen-api/blob/master/scripts/examples/python/XenAPI/XenAPI.py#L111 where a class that inherits `xmlrpc.client.Transport` redefines `send_request`.

`xmlrpc.client.Transport` doesn't have the same signature for `send_request` in python2 and python3, so a parameter is missing when called in python3. The backtrace is somewhat misleading as no mention of this function is made at all.

A similar issue is discussed at https://www.reddit.com/r/learnpython/comments/1l38mf/i_cant_get_xmlrpc_on_python_3_to_use_an_http_proxy/

stormi committed:
python 2 to 3: replace StringIO with io.BytesIO

futurize suggested using io.StringIO in python 2, but what tarfile
actually expects is a file in bytes mode. Use BytesIO instead and import
bytes from future's builtins for python2 compatibility.

Signed-off-by: Samuel Verschelde <stormi-xcp@ylix.fr>

stormi committed:
python 2 to 3: remove useless imports added by futurize

Signed-off-by: Samuel Verschelde <stormi-xcp@ylix.fr>

stormi commented:
Still WIP as I can't test it with XenAPI.py + python3 yet (see above).

I would also be glad to completely remove python2 support right from the start if it were possible, as retaining it first causes somewhat unneeded complexity in my view (though now that it's done, it doesn't represent a lot of code... See the last commit). I think the changes are simple enough that we could squash everything into a single "port to python 3" commit.

stormi commented:
I've tested the penultimate commit with python2 on a 8.2 host, exported the templates parameters with the following python script, and compared with what there was on the host before. Only `OpaqueRef`s and `mac_seed`s changed, as expected.

```python
#!/usr/bin/env python3

import subprocess

output = subprocess.check_output("xe template-list params=uuid,name-label", shell=True).decode()

templates = []
for line in output.splitlines():
    if line.startswith('uuid'):
        uuid = line.split(':')[1].strip()
    if line.strip().startswith('name-label'):
        label = line.split(':')[1].strip()
        templates.append((label, uuid))

for name, uuid in sorted(templates):
    subprocess.check_call(f"xe template-param-list uuid={uuid}", shell=True)
```

edwintorok commented:
XenAPI.py shouldn't override `send_request`, it should override `make_connection` instead, see https://stackoverflow.com/questions/11729159/use-python-xmlrpclib-with-unix-domain-sockets for some ideas.
Probably best to address that in a PR of its own, that'll make XenAPI.py work better with py2/3 (you can test it by trying to make an xmlrpc connection over the local Unix socket in Dom0)

stormi commented:
Does anyone know why `send_request` was overridden in the first place?

stormi commented:
The PR is now ready for review.

With XenAPI.py now fixed, I could finalize the port to python 3. The main change is that StringIO was replaced with BytesIO, as tarfile expects a bytes-like object (see commit https://github.com/xenserver/guest-templates-json/pull/1/commits/39b81ac5d8fb0bfafc5fc31cc6487dc085828fb5).

On the penultimate commit, before the one that removes python2 compatibility, I ran `check.py` and did the test that I described yesterday, with python2.

On the last commit, I did the same with python 3.6. There is a slight change in the output in this case: the order of items in the `platform` parameter changed. Example:
```diff
- platform (MRW): videoram: 8; hpet: true; secureboot: false; device-model: qemu-upstream-compat; apic: true; device_id: 0001; vga: std; nx: true; pae: true; viridian: false; acpi: 1
+ platform (MRW): vga: std; videoram: 8; viridian: false; device_id: 0001; device-model: qemu-upstream-compat; secureboot: false; nx: true; acpi: 1; apic: true; pae: true; hpet: true
```

The number of characters is exactly the same, as well as the contents, as far as I could tell with a visual check on the templates I verified.

stormi commented:
Last push was just to fix a typo in a commit message.

stormi commented:
Now that python3 is available in the platform, it might be the right time to finalize this PR?

See comment https://github.com/xenserver/guest-templates-json/pull/1#issuecomment-1199791614
